homeassistant/components/atome/sensor.py

```diff
@@ -0,0 +1,319 @@
+"""Linky Atome."""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+from pyatome.client import AtomeClient
+from pyatome.client import PyAtomeError
+
+from homeassistant.const import (
+    CONF_PASSWORD,
+    CONF_USERNAME,
+    CONF_NAME,
+    DEVICE_CLASS_POWER,
+    POWER_WATT,
+    ENERGY_KILO_WATT_HOUR,
+)
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "atome"
+
+SESSION_RENEW_INTERVAL = timedelta(minutes=55)
+
+LIVE_SCAN_INTERVAL = timedelta(seconds=30)
+DAILY_SCAN_INTERVAL = timedelta(seconds=150)
+WEEKLY_SCAN_INTERVAL = timedelta(hours=1)
+MONTHLY_SCAN_INTERVAL = timedelta(hours=1)
+YEARLY_SCAN_INTERVAL = timedelta(days=1)
+
+LIVE_NAME = "Atome Live Power"
+DAILY_NAME = "Atome Daily"
+WEEKLY_NAME = "Atome Weekly"
+MONTHLY_NAME = "Atome Monthly"
+YEARLY_NAME = "Atome Yearly"
+
+LIVE_TYPE = "live"
+DAILY_TYPE = "day"
+WEEKLY_TYPE = "week"
+MONTHLY_TYPE = "month"
+YEARLY_TYPE = "year"
+
+ICON = "mdi:flash"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Atome sensor."""
+    username = config[CONF_USERNAME]
+    password = config[CONF_PASSWORD]
+
+    data = AtomeData(username, password)
+
+    @Throttle(LIVE_SCAN_INTERVAL)
+    def update_live():
+        """Update the live power usage."""
+        data.get_live_power()
+
+    @Throttle(DAILY_SCAN_INTERVAL)
+    def update_daily():
+        """Update the daily power usage."""
+        data.get_daily_usage()
+
+    @Throttle(WEEKLY_SCAN_INTERVAL)
+    def update_weekly():
+        """Update the weekly power usage."""
+        data.get_weekly_usage()
+
+    @Throttle(MONTHLY_SCAN_INTERVAL)
+    def update_monthly():
+        """Update the monthly power usage."""
+        data.get_monthly_usage()
+
+    @Throttle(YEARLY_SCAN_INTERVAL)
+    def update_yearly():
+        """Update the yearly power usage."""
+        data.get_yearly_usage()
+
+    update_live()
+    update_daily()
+    update_weekly()
+    update_monthly()
+    update_yearly()
+
+    add_entities([AtomeSensor(data, LIVE_NAME, LIVE_TYPE, update_live)], True)
+    add_entities([AtomeSensor(data, DAILY_NAME, DAILY_TYPE, update_daily)], True)
+    add_entities([AtomeSensor(data, WEEKLY_NAME, WEEKLY_TYPE, update_weekly)], True)
+    add_entities([AtomeSensor(data, MONTHLY_NAME, MONTHLY_TYPE, update_monthly)], True)
+    add_entities([AtomeSensor(data, YEARLY_NAME, YEARLY_TYPE, update_yearly)], True)
```

Comment:
Collect all entities first and then make a single call to `add_entities`.

Reply:
ok, done